src/python/grpcio_tests/tests/unit/_dns_resolver_test.py

```diff
@@ -0,0 +1,58 @@
+# Copyright 2019 The gRPC Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for an actual dns resolution."""
+
+import unittest
+import logging
+import six
+
+import grpc
+from tests.unit import test_common
+
+_METHOD = '/ANY/METHOD'
+_REQUEST = b'\x00\x00\x00'
+_RESPONSE = _REQUEST
+
+
+class GenericHandler(grpc.GenericRpcHandler):
+
+    def service(self, unused_handler_details):
+        return grpc.unary_unary_rpc_method_handler(
+            lambda request, unused_context: request,
+        )
+
+
+class DNSResolverTest(unittest.TestCase):
+
+    def setUp(self):
+        self._server = test_common.test_server()
+        self._server.add_generic_rpc_handlers((GenericHandler(),))
+        self._port = self._server.add_insecure_port('[::]:0')
+        self._server.start()
+
+    def tearDown(self):
+        self._server.stop(None)
+
+    def test_connect_loopback(self):
+        # NOTE(https://github.com/grpc/grpc/issues/18422)
+        # In short, Gevent + C-Ares = Segfault. The C-Ares driver is not
+        # supported by custom io manager like "gevent" or "libuv".
+        with grpc.insecure_channel(
+                'loopback4.unittest.grpc.io:%d' % self._port) as channel:
```

Comment:
ipv6 should be available even in tests that run in docker containers on Kokoro.

Was the ipv6 availability issue encountered while running this test manually?

Reply:
A bit more on that, IMO tests should assume that ipv6 is available, but not that ipv4 is available, because all grpc unit tests run on either dual stack or ipv6-only environments. Tests depending explicitly on ipv4 prevents such tests from ever running in ipv6-only environments, and futher keeps around the need for dual-stack.

Reply:
The `IPv6` fail log is caused by the second commit, here is the full log: https://source.cloud.google.com/results/invocations/3799a132-06ea-494c-91c4-6466f18fa304/targets/grpc%2Fcore%2Fpull_request%2Flinux%2Fgrpc_python_bazel_test/log

Here is the highlight:
```
======================================================================
ERROR: test_connect_loopback (__main__.DNSResolverTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/.cache/bazel/_bazel_root/d2dc70c3d9da3fab488ba0dcbbd35051/execroot/com_github_grpc_grpc/bazel-out/k8-fastbuild/bin/src/python/grpcio_tests/tests/unit/_dns_resolver_test.runfiles/com_github_grpc_grpc/src/python/grpcio_tests/tests/unit/_dns_resolver_test.py", line 53, in test_connect_loopback
    self.assertEqual(channel.unary_unary(_METHOD)(_REQUEST), _RESPONSE)
  File "/root/.cache/bazel/_bazel_root/d2dc70c3d9da3fab488ba0dcbbd35051/execroot/com_github_grpc_grpc/bazel-out/k8-fastbuild/bin/src/python/grpcio_tests/tests/unit/_dns_resolver_test.runfiles/com_github_grpc_grpc/src/python/grpcio/grpc/_channel.py", line 562, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/root/.cache/bazel/_bazel_root/d2dc70c3d9da3fab488ba0dcbbd35051/execroot/com_github_grpc_grpc/bazel-out/k8-fastbuild/bin/src/python/grpcio_tests/tests/unit/_dns_resolver_test.runfiles/com_github_grpc_grpc/src/python/grpcio/grpc/_channel.py", line 466, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
_Rendezvous: <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "DNS resolution failed"
	debug_error_string = "{"created":"@1553714241.572645062","description":"Failed to create subchannel","file":"src/core/ext/filters/client_channel/client_channel.cc","file_line":2715,"referenced_errors":[{"created":"@1553714241.572633454","description":"Resolver transient failure","file":"src/core/ext/filters/client_channel/resolving_lb_policy.cc","file_line":335,"referenced_errors":[{"created":"@1553714241.572629488","description":"DNS resolution failed","file":"src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc","file_line":327,"grpc_status":14,"referenced_errors":[{"created":"@1553714241.572531047","description":"C-ares status is not ARES_SUCCESS: DNS server returned answer with no data","file":"src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc","file_line":240}]}]}]}"
>
```

What do you think is the best solution to satisfy both  Bazel in Kokoro/Docker and IPv6 only environment?

Reply:
Hmm, yeah I guess these tests won't be able to use ipv6. In that case, use of the ipv4 loopback is<i>ok</i>

FWIW I checked the kokoro configs and it looks like the problem is that, unlike tests that use `build_docker_and_run_tests.sh` to invoke `docker run`, that bazel tests uses `build_and_run_docker.sh` to invoke `docker run` - and the latter doesn't supply the `--sysctl net.ipv6.conf.all.disable_ipv6=0` flag. It would probably be best to update `build_docker_and_run_tests.sh` to support ipv6 but I think that doesn't need to be done in this PR - cc @jtattermusch